### DISCUSSION TITLE
parse JSON

### DISCUSSION THREAD

**Discussion by simonarbel** [grafana/grafana-infinity-datasource#547] [Q&A]
Hi
What is needed to be changed in the following UQL in order to catch data metric for all FWs?

```kusto 
parse-json
| scope "value"
| project "Name"="id","Total Bytes"=sum("timeseries[0].data[0].total")
```

Body:
```json 
{
    "cost": 239,
    "timespan": "2023-02-23T01:00:00Z/2023-02-23T05:00:00Z",
    "interval": "P1D",
    "value": [
        {
            "id": "/subscriptions/XXXXXXX/resourceGroups/DMZ-Network-RG/providers/Microsoft.Network/azureFirewalls/DMZ-Firewall/providers/Microsoft.Insights/metrics/DataProcessed",
            "type": "Microsoft.Insights/metrics",
            "name": {
                "value": "DataProcessed",
                "localizedValue": "Data processed"
            },
            "displayDescription": "Total amount of data processed by this firewall",
            "unit": "Bytes",
            "timeseries": [
                {
                    "metadatavalues": [],
                    "data": [
                        {
                            "timeStamp": "2023-02-23T01:00:00Z",
                            "total": 19610734178
                        }
                    ]
                }
            ],
            "errorCode": "Success"
        }
    ],
    "namespace": "Microsoft.Network/azureFirewalls",
    "resourceregion": "westeurope"
}
```

**Comment by yesoreyeram** [comment]
What do you mean "data metric for all FWs?" your sample data looks like having one one FW.

Can you post the sample response json and sample output what you are expecting? 

BTW, adding more items to the value array just gives me this without any issues.

<img width="1790" alt="image" src="https://user-images.githubusercontent.com/153843/223076239-9c420cf4-7802-4025-b46c-a4b4b33a3145.png">

**Comment by simonarbel** [comment]
Cannot find an example output of more than one FW. But theoretically,
what would be the correct UQL syntax if there was one more item in the “value” array:

"value": [
        {
            "id": "/subscriptions/5423e28d-af1f-4946-ab18-1d0c2dc0224a/resourceGroups/DMZ-Network-RG/providers/Microsoft.Network/azureFirewalls/DMZ-Firewall/providers/Microsoft.Insights/metrics/DataProcessed",
            "type": "Microsoft.Insights/metrics",
            "name": {
                "value": "DataProcessed",
                "localizedValue": "Data processed"
            },
            "displayDescription": "Total amount of data processed by this firewall",
            "unit": "Bytes",
            "timeseries": [
                {
                    "metadatavalues": [],
                    "data": [
                        {
                            "timeStamp": "2023-02-23T01:00:00Z",
                            "total": 19610734178
                        }
                    ]
                }
            ],

**Comment by simonarbel** [comment]
I’m trying to use [*]  but it refuses to give me results…

From: Simon Arbel
Sent: Monday, March 6, 2023 12:13 PM
To: yesoreyeram/grafana-infinity-datasource ***@***.***>; yesoreyeram/grafana-infinity-datasource ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [yesoreyeram/grafana-infinity-datasource] parse JSON (Discussion #547)

Cannot find an example output of more than one FW. But theoretically,
what would be the correct UQL syntax if there was one more item in the “value” array:

"value": [
        {
            "id": "/subscriptions/5423e28d-af1f-4946-ab18-1d0c2dc0224a/resourceGroups/DMZ-Network-RG/providers/Microsoft.Network/azureFirewalls/DMZ-Firewall/providers/Microsoft.Insights/metrics/DataProcessed",
            "type": "Microsoft.Insights/metrics",
            "name": {
                "value": "DataProcessed",
                "localizedValue": "Data processed"
            },
            "displayDescription": "Total amount of data processed by this firewall",
            "unit": "Bytes",
            "timeseries": [
                {
                    "metadatavalues": [],
                    "data": [
                        {
                            "timeStamp": "2023-02-23T01:00:00Z",
                            "total": 19610734178
                        }
                    ]
                }
            ],

**Reply by yesoreyeram** [reply]
`*` is not a valid syntax in this context. What you need is `mv-expand` command. mv-expand works same way as [kusto](https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/mvexpandoperator).

**Comment by simonarbel** [comment]
***@***.***

From: Simon Arbel
Sent: Monday, March 6, 2023 12:25 PM
To: 'yesoreyeram/grafana-infinity-datasource' ***@***.***>; 'yesoreyeram/grafana-infinity-datasource' ***@***.***>
Cc: 'Author' ***@***.***>
Subject: RE: [yesoreyeram/grafana-infinity-datasource] parse JSON (Discussion #547)

I’m trying to use [*]  but it refuses to give me results…

From: Simon Arbel
Sent: Monday, March 6, 2023 12:13 PM
To: yesoreyeram/grafana-infinity-datasource ***@***.******@***.***>>; yesoreyeram/grafana-infinity-datasource ***@***.******@***.***>>
Cc: Author ***@***.******@***.***>>
Subject: RE: [yesoreyeram/grafana-infinity-datasource] parse JSON (Discussion #547)

Cannot find an example output of more than one FW. But theoretically,
what would be the correct UQL syntax if there was one more item in the “value” array:

"value": [
        {
            "id": "/subscriptions/5423e28d-af1f-4946-ab18-1d0c2dc0224a/resourceGroups/DMZ-Network-RG/providers/Microsoft.Network/azureFirewalls/DMZ-Firewall/providers/Microsoft.Insights/metrics/DataProcessed",
            "type": "Microsoft.Insights/metrics",
            "name": {
                "value": "DataProcessed",
                "localizedValue": "Data processed"
            },
            "displayDescription": "Total amount of data processed by this firewall",
            "unit": "Bytes",
            "timeseries": [
                {
                    "metadatavalues": [],
                    "data": [
                        {
                            "timeStamp": "2023-02-23T01:00:00Z",
                            "total": 19610734178
                        }
                    ]
                }
            ],

**Comment by simonarbel** [comment]
Can you provide the fixed UQL syntax please?

**Comment by yesoreyeram** [comment]
I can't give you step by step detail..

but the following may give you some idea

```
parse-json
| scope "value"
| mv-expand "series"="timeseries[0].data"
| project "name"="id", "ts"="series.timeStamp", "value"="series.total"
```

**Comment by simonarbel** [comment]
Thanks this one works 😊

**Comment by simonarbel** [comment]
HI

I’m trying to use the UQL extract command to be able to display only part of the string using Regex:

/subscriptions/69d34344-d7b7-4dc2-aefa-fda3c77fb570/resourceGroups/PAW/providers/Microsoft.Network/azureFirewalls/AzureFirewall_VWAN-HUB/providers/Microsoft.Insights/metrics/DataProcessed

Need to display only the part in red.

This is the code:
parse-json
| scope "value"
| mv-expand "items"="timeseries[0].data"
| project "Name"=extract('.*/(.*)',index,"id"),"Total Bytes"="items.total"


But it gives me empty value. What is the correct syntax?




From: Simon Arbel
Sent: Monday, March 6, 2023 2:29 PM
To: yesoreyeram/grafana-infinity-datasource ***@***.***>; yesoreyeram/grafana-infinity-datasource ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [yesoreyeram/grafana-infinity-datasource] parse JSON (Discussion #547)

Thanks this one works 😊

**Comment by yosiasz** [comment]
check this out

https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/558
